### DISCUSSION TITLE
Error Missing permission. Redirect to login?

### DISCUSSION THREAD

**Discussion by schwantuschke** [kimai/kimai#4797] [On-Premise: Installation & Updates] [answered]
### Describe the issue

If I want to Restart one of your last activities or I create a new activity and change Customer, Project or Activity, I get the error message _Missing permission. Redirect to login?_

If I choose Confirm, I get to the dashboard.
There are no current entries in the log file.

### I already tried

- [X] I've read and searched [the documentation](https://www.kimai.org/documentation/).
- [X] I've searched for similar issues in this repository.
- [X] I've searched for similar issues in [the discussions](https://github.com/kimai/kimai/discussions).

### Kimai version

2.15.0

### How do you run Kimai?

Virtual Server or alike

### Which PHP version are you using?

8.1

### Screenshots

<img width="392" alt="Bildschirmfoto 2024-04-17 um 15 32 47" src="https://github.com/kimai/kimai/assets/4090141/75be5cac-7847-4a28-81eb-a2f2d3d231e4">

**Comment by kevinpapst** [comment]
That can have a lot of reasons. Likely an invalid set of role permissions or a broken Apache config.
Check what the network tab reports on the API calls, when you change the customer/project

**Reply by schwantuschke** [reply]
I get an error 403. I can't get any further here and will contact my hoster. Thank you

**Reply by kevinpapst** [reply]
Did this happen after the update to 2.15? 
There have been changes to the Apache .htaccess in public/ which are necessary for some environments.

**Reply by schwantuschke** [reply]
Yes, it hasn't been faulty for that long. It's possible that it came with the update.

**Reply by kevinpapst** [reply]
Make sure these rules for the `HTTP_AUTHORIZATION` header are in place, for some weird reason Apache does not forward them to PHP by default: https://github.com/kimai/kimai/blob/main/public/.htaccess#L40

**Reply by schwantuschke** [reply]
The rule are in place…

**Reply by kevinpapst** [reply]
Your error message is vey much the same:
https://github.com/kimai/kimai/discussions/4750
https://github.com/kimai/kimai/issues/4774

**Comment by dani** [comment]
I have the same error since 2.14.0 (and also on 2.15.0). Adding
```
SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1
```
makes no difference. I'm running PHP 8.3 as FPM, with httpd + mod_proxy_fcgi. This config was working since a year and a half through various Kimai versions. Using either local auth, or SAML.

**Reply by kevinpapst** [reply]
I am using nginx everywhere. I didn't have to change anything. 
https://symfony.com/doc/current/setup/web_server_configuration.html#nginx

**Reply by dani** [reply]
Switching to nginx could possibly be a workarround (but not a trivial one though, as I deploy Kimai using some common ansible roles, also used by other apps). But I'd prefer understand what's going on since 2.14.0

**Reply by dani** [reply]
BTW, it seems to be only /api which is affected. For example, I can create an API token from my profile page (which sends a POST on /fr_FR/profile/user@foo.bar/create-access-token) but then cannot delete it, as the DELETE is sent on /api/users/api-token/2, for which Kimai replies with a 403

**Reply by kevinpapst** [reply]
Can you show me a screenshot of your network console - which headers are sent => the full request. And then the full response as well. There is a new API authentication system in place. It should only take over if you send a proper API call.

Please try if the regular API flow works, when you use a new API token, e.g. via /api/doc/ 

And you can start debugging here: 
https://github.com/kimai/kimai/blob/main/src/API/Authentication/ApiRequestMatcher.php
Check if your request from the UI is matched or not (it shouldn't, as you are already logged-in).

**Reply by dani** [reply]
![Capture d’écran du 2024-04-29 10-42-29](https://github.com/kimai/kimai/assets/526826/a4767da1-b813-4366-8886-9faab3f45a81)

Here's a req to the API which is denied.

In the ApiRequestMatcher, it looks like it's matched by
```php
        // let's use this firewall if a Bearer token is set in the header
        if ($request->headers->has('Authorization')) {
            return true;
        }
```

Commenting this, I still get a 403, but a different one (message in the response payload is Forbidden, while it's "Access denied" before I commented the code snippet).

**Reply by kevinpapst** [reply]
Can you lo-out and log-in and try the same again. Seems you have a RememberMe Cookie, but no real session ID.

**Reply by dani** [reply]
I'll try to debug API access manually, but I don't even have access to /api/doc (403 Forbiden here too)

**Reply by dani** [reply]
Tried to logout, manually removed the REMEMBERME cookie, login again. No change. So, it seems there's a issue with sessions, I'll try to look in this area

**Reply by kevinpapst** [reply]
Which cookies did you have after your re-login?
And how did you login? User/PW or SAML? This SSO Cookie is interesting...

**Reply by dani** [reply]
It was with saml

**Reply by dani** [reply]
After a re-login, I have the following cookies
![image](https://github.com/kimai/kimai/assets/526826/43f571b7-9941-44d6-9e71-1aa256e10d4e)

(there's also a 4th one, which is unrelated as it's my SSO cookie)

**Reply by kevinpapst** [reply]
And the request still fails, even with PHPSESSID cookie?
Is the PHPSESSID cookie sent when you execute an API call, eg restart recent activity?

**Reply by dani** [reply]
Yes, the PHPSESSID cookie is sent in the request, and the request still fails (it was already set and sent before the logout/login). The session is correctly inserted in kimai2_sessions table too. With this content
```
_sf2_attributes|a:4:{s:18:"_csrf/https-logout";s:43:"LBO7xQN0-TiDGQHUFwDaifZ7cE6me7ros4DSof6_tio";s:18:"_csrf/https-search";s:43:"SxnihiGDe8nRtXUxXytT_uhWHG0Yfu9eXf044MRAxO8";s:28:"_csrf/https-datatable_update";s:43:"TokCiNi1tncC1tFlpEgrkhnrcFm4VxLlE-D69SRFsT0";s:22:"_security_secured_area";s:417:"O:68:"Symfony\Component\Security\Core\Authentication\Token\RememberMeToken":3:{i:0;s:50:"4{gjsH=]l;5NtAT(r$/+dai6%Bk4;71n?Zn?a2T*1fwLSUmiH?";i:1;s:12:"secured_area";i:2;a:5:{i:0;O:15:"App\Entity\User":5:{s:2:"id";i:4;s:8:"username";s:16:"dani@***.***";s:7:"enabled";b:1;s:5:"email";s:16:"dani@***.***";s:8:"password";s:0:"";}i:1;b:1;i:2;N;i:3;a:0:{}i:4;a:2:{i:0;s:16:"ROLE_SUPER_ADMIN";i:1;s:9:"ROLE_USER";}}}";}_sf2_meta|a:3:{s:1:"u";i:1714382839;s:1:"c";i:1714382836;s:1:"l";i:0;}
```

**Reply by kevinpapst** [reply]
Lets get back to your answers a few steps before. 
You said that this `if ($request->headers->has('Authorization')) {` matches, correct? 
But the screenshot has not such header, so can you add a `var_dump($request->headers->get('Authorization'));exit;` at that place?

Maybe something (like a global proxy/loadbalancer/sso solution) injects that header empty or with a value that is not meant for Kimai?

**Reply by dani** [reply]
Yes, it's matching. Somehow, Authorization is sent as an empty string, output with the var_dump statement is
```
string(0) ""
```
I do have a global nginx reverse proxy, but capturing the trafic received by apache (so, after the reverse proxy) shows no Authorization header being added

**Reply by kevinpapst** [reply]
I will add a check for non empty header, even though that sounds more like an environment issue. 
Headers that suddenly appear out of the blue are not ideal

**Reply by dani** [reply]
Adding the non empty string check, like this
```php
        // let's use this firewall if a Bearer token is set in the header
        if ($request->headers->has('Authorization') && $request->headers->get('Authorization') != "") {
            return true;
        }
```
makes the api working again. I'll try to investigate where this empty header comes from

**Reply by dani** [reply]
Many thanks for your time, and help with debuging this !

**Reply by kevinpapst** [reply]
You are not the only one with that issue, so it seems to be a somewhat "common issue" that depends on the used infrastructure. Which is always the worst to figure out, if you can't see it yourself.

Thanks for helping 👍 

If you find the reason for the header, would be great to share it for the next person

**Comment by kevinpapst** [comment]
Ok, I can reproduce the behavior if I only have a REMEMBERME Cookie, but no PHPSESSID.
But usually you should immediately get a new PHPSESSID if you only come with a REMEMBERME cookie.
I'll need to investigate

**Comment by schwantuschke** [comment]
An update to the latest version 2.16.0 fixed the error.

**Reply by kevinpapst** [reply, marked as answer]
Hooray 🎉  then it was the empty Authorization header. Thanks for the feedback!

**Reply by schwantuschke** [reply]
Thanks for fix 😀

**Reply by kevinpapst** [reply]
I am super glad we found the issue. I would never have found that on my own, so every input and help is always appreciated!
